### PR TITLE
[LTC] Upstream class Shape

### PR DESCRIPTION
**Pull request by alanwaketan**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #67672

Summary:
This commit Upstreams class Shape from lazy_tensor_staging branch.

Test Plan:
./build/bin/test_lazy --gtest_filter=ShapeTest.*

Differential Revision: [D32095478](https://our.internmc.facebook.com/intern/diff/D32095478)

### CODE REVIEW

**Comment by alanwaketan** [comment]
@alanwaketan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32095478).

**Comment by desertfire** [comment]
I thought yesterday's discussion was leaning towards getting rid of Shape? I am not opposing keeping Shape, and in fact I like it because that will make our merging easier. Just want to clarify so that we are all on the same page.

**Comment by wconstab** [comment]
> I thought yesterday's discussion was leaning towards getting rid of Shape? I am not opposing keeping Shape, and in fact I like it because that will make our merging easier. Just want to clarify so that we are all on the same page.

No, we agreed to keep Shape yesterday, mainly for ease of having an operator== implementation.  But it'll still be a greatly simplified Shape (already removed most of the objectionable parts from it)

File: torch/csrc/lazy/core/shape.h [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    // should be for a tuple shape.  However, for tuple shapes, we appear
+    // to be accessing the element_type field in some places.  Fix this.
+    at_element_type_ = element_shapes[0].at_element_type();
+  }
+
+  std::string ToString(bool print_layout = false) const {
+    return c10::str(toString(at_element_type_), "[", c10::Join(",", dimensions_), "]");
+  }
+
+  c10::ScalarType at_element_type() const { return at_element_type_; }
```
Comment (wconstab) [inline torch/csrc/lazy/core/shape.h:33]: should we rename this to `scalar_type()` to be consistent?

File: torch/csrc/lazy/core/shape.h [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+  void set_element_type(at::ScalarType value);
+
+  int64_t rank() const { return dimensions_.size(); }
```
Comment (wconstab) [inline torch/csrc/lazy/core/shape.h:36]: and change this to `dim()` similarly?

File: torch/csrc/lazy/core/shape.h [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+#include <atomic>
+#include <ostream>
+#include <vector>
+
+#include <c10/core/Scalar.h>
+#include <c10/util/Logging.h>
+
+namespace lazy_tensors {
+
```
Comment (wconstab) [inline torch/csrc/lazy/core/shape.h:11]: not sure if you saw my message somewhere else, but can you also put the ShapeIndex class (From shape_util.h) here?  I moved it into shape.h in my big PR but that hasn't landed to staging yet.  ShapeIndex will be a dep of lowering_context unless we refactor that, but it's also pretty small and unoffensive looking

**Comment by alanwaketan** [comment]
@wconstab This question just came out of my mind. Is Shape supposed to be backend specific?

**Comment by alanwaketan** [comment]
> > @wconstab This question just came out of my mind. Is Shape supposed to be backend specific?
> 
> Trick question, I think.
> 
> lazy::Shape is not supposed to be backend specific. Meaning, it's designed to be used (a) by LTC code, in a backend-agnostic way, where it mirrors aten semantics faithfully; (b) as a way to pass aten shape information into backends
> 
> Backends may then take lazy::Shape and convert to e.g. xla::Shape (or subclass). At this time, they may add backend-specific information, but it shouldn't leak back out into the LTC APIs.

Then maybe we should have moved some of the shape related APIs from TsNode back to Node. That should be a separate task though.

The reason why I ask is more on whether the current directory is the right one to place the `Shape` implementation.

**Comment by wconstab** [comment]
> > > @wconstab This question just came out of my mind. Is Shape supposed to be backend specific?
> > 
> > 
> > Trick question, I think.
> > lazy::Shape is not supposed to be backend specific. Meaning, it's designed to be used (a) by LTC code, in a backend-agnostic way, where it mirrors aten semantics faithfully; (b) as a way to pass aten shape information into backends
> > Backends may then take lazy::Shape and convert to e.g. xla::Shape (or subclass). At this time, they may add backend-specific information, but it shouldn't leak back out into the LTC APIs.
> 
> Then maybe we should have moved some of the shape related APIs from TsNode back to Node. That should be a separate task though.
> 
> The reason why I ask is more on whether the current directory is the right one to place the `Shape` implementation.

Yea, we might end up putting Shape back into Node; I'm not sure if we need to or not.  

But I think it's right to put shape.h in the main folder not the backend folder, if that's your question.

**Comment by alanwaketan** [comment]
@alanwaketan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32095478).

**Comment by alanwaketan** [comment]
@alanwaketan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32095478).

**Review by wconstab — APPROVED** [review]

**Comment by alanwaketan** [comment]
Thanks @wconstab for approving this pull request. Waiting for the CI to land.

**Pull request closed without merging**